### DISCUSSION TITLE
how to achieve the Guidance acceleration

### DISCUSSION THREAD

**Discussion by xujli** [guidance-ai/guidance#515] [Q&A] [answered]
I just want to know how do you achieve the Guidance acceleration in transformers, only cache? Could you point out the source code I'd like to appreciate it.

**Comment by slundberg** [comment, marked as answer]
It works by sending things in batch to the compute backend when we can. How this batching happens is specific to the backend but the if statement that detect if the next token is forced (and can so be batched) is at: https://github.com/guidance-ai/guidance/blob/26554b05619347be666647a2ac97a6a51f467bfd/guidance/models/_model.py#L731

hope that helps!

**Reply by xujli** [reply]
If the next token is forced, the backend will directly return the token ID instead of get_logits, thanks for your reply. Thus, the actual acceleration is the parallel attention score calculation when we feed multiple new tokens caused by the template and avoid the logits processing for the template tokens, Am I right?
